src/transformers/models/megatron_bert/convert_megatron_bert_checkpoint.py

```diff
@@ -0,0 +1,268 @@
+
+    # The transformer.
+    transformer = lm["transformer"]
+
+    # The regex to extract layer names.
+    layer_re = re.compile("layers\.(\d+)\.([a-z0-9_.]+)\.([a-z]+)")
+
+    # The simple map of names for "automated" rules.
+    megatron_to_transformers = {
+        "attention.dense": ".attention.output.dense.",
+        "mlp.dense_h_to_4h": ".intermediate.dense.",
+        "mlp.dense_4h_to_h": ".output.dense.",
+    }
+
+    # Keep track of the attention/query/value tensor.
+    attention_qkv_weight = None
+
+    # Extract the layers.
+    for key, val in transformer.items():
+        # Match the name.
+        m = layer_re.match(key)
+
+        # Stop if that's not a layer
+        if m is None:
+            break
+
+        # The index of the layer.
+        layer_idx = int(m.group(1))
+        # The name of the operation.
+        op_name = m.group(2)
+        # Is it a weight or a bias?
+        weight_or_bias = m.group(3)
+
+        # The name of the layer.
+        layer_name = "bert.encoder.layer.{}".format(layer_idx)
+
+        # For layernorm(s), simply store the layer norm.
+        if op_name.endswith("layernorm"):
+
+            ln_name = "attention.ln" if op_name.startswith("input") else "ln"
+            output_state_dict[layer_name + "." + ln_name + "." + weight_or_bias] = val
+
+        # Transpose the QKV matrix.
+        elif op_name == "attention.query_key_value" and weight_or_bias == "weight":
+
+            # Make sure the QKV pointer is nil.
+            assert attention_qkv_weight is None, ""
+
+            # Store the tensor as we need the bias as well to interleave QKV and biases.
+            attention_qkv_weight = val
+
+        # Transpose the bias.
+        elif op_name == "attention.query_key_value" and weight_or_bias == "bias":
+
+            # Make sure we read the weight tensor.
+            assert attention_qkv_weight is not None, ""
+
+            # Split the QKV matrix into Q, K and V. Megatron stores Q,K,V interleaved.
+            q = attention_qkv_weight[0 * 1024 : 1 * 1024, :]
+            k = attention_qkv_weight[1 * 1024 : 2 * 1024, :]
+            v = attention_qkv_weight[2 * 1024 : 3 * 1024, :]
+
+            # Split the bias.
+            q_bias = val[0 * 1024 : 1 * 1024]
+            k_bias = val[1 * 1024 : 2 * 1024]
+            v_bias = val[2 * 1024 : 3 * 1024]
+
+            # The name of the self attention block.
+            self_attn_name = layer_name + ".attention.self"
+
+            # Store.
+            output_state_dict[self_attn_name + ".query.weight"] = q
+            output_state_dict[self_attn_name + ".query.bias"] = q_bias
+            output_state_dict[self_attn_name + ".key.weight"] = k
+            output_state_dict[self_attn_name + ".key.bias"] = k_bias
+            output_state_dict[self_attn_name + ".value.weight"] = v
+            output_state_dict[self_attn_name + ".value.bias"] = v_bias
```

Comment:
Nit, the keys would be more readable as f-strings: `f"{self_attn_name}.query.weight"` for instance.

Reply:
Addressed in commit f155bc48ffd4606aab7461035e49c0318e57fd9c with `f"{layer_name}.attention.self.query.weight"` and friends.